packages/watch-eth/test/ethereum-event-processor.spec.ts

```diff
@@ -0,0 +1,123 @@
+import './setup'
+/* External Imports */
+import { getLogger, newInMemoryDB } from '@pigi/core'
+import { createMockProvider, getWallets } from 'ethereum-waffle'
+
+/* Internal Imports */
+import { deployTokenContract, TestListener } from './utils'
+import { Event, EthereumEventProcessor } from '../src'
+
+const log = getLogger('ethereum-event-processor', true)
+
+const timeout = 25_000
+describe('Event Subscription', () => {
+  let provider
+  let wallets
+  let ownerWallet
+  let recipientWallet
+
+  const sendAmount = 100
+  const initialSupply = 100_000
+
+  let tokenContract
+  let eventProcessor: EthereumEventProcessor
+  let eventListener: TestListener<Event>
+
+  beforeEach(async () => {
+    provider = createMockProvider()
+    wallets = getWallets(provider)
+    ownerWallet = wallets[0]
+    recipientWallet = wallets[1]
+
+    log.debug(`Connection info: ${JSON.stringify(provider.connection)}`)
+
+    tokenContract = await deployTokenContract(ownerWallet, initialSupply)
+
+    eventProcessor = new EthereumEventProcessor(newInMemoryDB())
+    eventListener = new TestListener<Event>()
+  })
+
+  it('deploys correctly', async () => {
+    const ownerBalance = +(await tokenContract.balanceOf(ownerWallet.address))
+    ownerBalance.should.equal(initialSupply)
+  })
+
+  it('processes new events', async () => {
+    await eventProcessor.subscribe(
+      tokenContract,
+      'Transfer',
+      eventListener,
+      false
+    )
+
+    await tokenContract.transfer(
+      ownerWallet.address,
+      recipientWallet.address,
+      sendAmount
+    )
+
+    const events = await eventListener.waitForReceive()
+    events.length.should.equal(1)
+    const event: Event = events[0]
+    event.values['from'].should.equal(ownerWallet.address)
+    event.values['to'].should.equal(recipientWallet.address)
+    event.values['amount'].toNumber().should.equal(sendAmount)
+  }).timeout(timeout)
+
+  it('processes old events', async () => {
```

Comment:
You said that this implementation also stores the events which have already been processed so if we want to we can skip through all of these events I presume?

Reply:
Yep, we store the last successfully processed block [here](https://github.com/plasma-group/pigi/pull/473/files#diff-4fdcbba76dbc4f9595f420a5666270bcR62) and use that as our `start` when we're backfilling things that we missed when we startup.

Reply:
Also can pass `false` in for the `backfill` parameter to `subscribe(...)` (defaults to `true`) or pass a start block number to the constructor.

Reply:
awesome!